model/message_broker.go

```diff
@@ -0,0 +1,135 @@
+// ForeignID is the id of a document located in another collection used by the
+// Topic interface to extract the actual message. The id of each message entry
+// is an int sequentially incremented for each message published to the topic.
+type MessageEntry struct {
+	ForeignID string `bson:"foreign_id"`
+
+	// ResumeToken and Error should be checked when watching a topic.
+	ResumeToken []byte
+	Err         error
+}
+
+// PublishToTopic creates a new message entry with the given id in the
+// corresponding collection for the given topic. If the topic does not exist in
+// the global registry, an error is returned.
+func PublishToTopic(ctx context.Context, env cedar.Environment, name string, id string) error {
+	topic, err := GetTopic(name)
+	if err != nil {
+		return err
+	}
+
+	return publishToTopic(ctx, env, topic, id)
+}
+
+// WatchTopic returns a MessageEntry channel fed by a change stream on the
+// topic's collection. The resume token specifies where to start after with the
+// change stream. If the topic does not exist in the global registry, an error
+// is returned.
+func WatchTopic(ctx context.Context, env cedar.Environment, name string, resumeToken []byte) (chan MessageEntry, error) {
+	topic, err := GetTopic(name)
+	if err != nil {
+		return nil, err
+	}
+
+	return watchTopic(ctx, env, topic, resumeToken)
+}
+
+// publishToTopic creates a new message entry with the foreign ID in the
+// corresponding collection for the given topic.
+func publishToTopic(ctx context.Context, env cedar.Environment, topic Topic, foreignID string) error {
+	entry := &MessageEntry{ForeignID: foreignID}
+	// TODO: should there be a collection per topic? Or, rather, one
+	// collection for all topics, indexed by topic names.
+	insertResult, err := env.GetDB().Collection(topic.Name()).InsertOne(ctx, entry)
+	grip.DebugWhen(err == nil, message.Fields{
+		"collection":   topic.Name(),
+		"insertResult": insertResult,
+		"op":           "publish to a topic",
+	})
+
+	return errors.Wrapf(err, "problem publishing to topic %s", topic.Name())
+}
+
+func watchTopic(ctx context.Context, env cedar.Environment, topic Topic, resumeToken []byte) (chan MessageEntry, error) {
+	// TODO: Maybe need to set options FullDocument -> options.UpdateLookup.
+	opts := options.ChangeStream()
+	if len(resumeToken) > 0 {
+		opts = opts.SetResumeAfter(resumeToken)
+	}
+	cs, err := env.GetDB().Collection(topic.Name()).Watch(ctx, mongo.Pipeline{}, opts)
+	if err != nil {
+		return nil, errors.Wrapf(err, "problem getting change streams for topic %s", topic.Name())
+	}
+
+	// TODO: figure out appropriate buffer size.
+	data := make(chan MessageEntry, 100)
+	go func() {
+		defer close(data)
+
+		for cs.Next(ctx) {
+			resumeToken = []byte(cs.ResumeToken())
+			entry := MessageEntry{}
+			entry.ResumeToken = resumeToken
+
+			if err = cs.Decode(cs); err != nil {
+				entry.Err = errors.Wrapf(err, "problem decoding message entry for topic %s", topic.Name())
+				data <- entry
```

Comment:
`cs.Next` checks for `ctx.Done()` so it will return `false` when that happens.

Reply:
It does, but we still have to ensure that data won't hang when it sends to the channel. If the loop successfully gets the next change and then the context cancels, then the outer function (in this case `Subscribe()`) will return with an error when sending the message to the client because the context was cancelled. The data stream won't close until it hits the deferred close, so the goroutine will be stuck sending to a channel that will never receive.

Reply:
Ah right, changing it now.